### ISSUE TITLE
Method for Full Manual Mode for Sensor OV7725

### ISSUE THREAD

**Issue by ajacobs-github** [openmv/openmv#887]
For full manual control of the OV7725 sensor, you need to disable auto gain, auto white balance, auto exposure, night mode (auto frame rate) and black level calibration (BLC). Only AFTER all of these auto modes are disabled, restore a bunch of offset registers back to their default values, because these get changed based on the initial images by the various auto modes before they are disabled.

Also manually set white balance.
```Python
# Setup Image Sensor format
sensor.set_framesize(sensor.QVGA)
sensor.set_pixformat(sensor.GRAYSCALE)

# Setup for manual gain, exposure, and white balance
sensor.set_auto_gain(False, gain_db=3) # gain in db
sensor.set_auto_whitebal(False)
sensor.set_auto_exposure(False, exposure_us=10000) # exposure in us

# Setup contrast, brightness and saturation
sensor.set_contrast(3) # range -3 to +3
sensor.set_brightness(0) # range -3 to +3
sensor.set_saturation(0) # range -3 to +3

# Disable night mode (auto frame rate) and black level calibration (BLC)
sensor.__write_reg(0x0E, 0b00000000) # Disable night mode
sensor.__write_reg(0x3E, 0b00000000) # Disable BLC

# Set White Balance values manually
sensor.__write_reg(0x01, 0b01111100) # Blue Gain for White Balance
sensor.__write_reg(0x02, 0b01000000) # Red Gain for White Balance
sensor.__write_reg(0x03, 0b01111100) # Green Gain for White Balance

# Restore to default values (these are changed before manual mode takes effect)
sensor.__write_reg(0x2D, 0b00000000) # LSB Insert Dummy Rows (Set to Default 0x00)
sensor.__write_reg(0x2E, 0b00000000) # MSB Insert Dummy Rows (Set to Default 0x00)

sensor.__write_reg(0x35, 0b10000000) # AD Offset B Chan (Set to Default 0x80)
sensor.__write_reg(0x36, 0b10000000) # AD Offset R Chan (Set to Default 0x80)
sensor.__write_reg(0x37, 0b10000000) # AD Offset Gb Chan (Set to Default 0x80)
sensor.__write_reg(0x38, 0b10000000) # AD Offset Gr Chan (Set to Default 0x80)

sensor.__write_reg(0x39, 0b10000000) # B channel offset (Set to Default 0x80)
sensor.__write_reg(0x3A, 0b10000000) # R channel offset (Set to Default 0x80)
sensor.__write_reg(0x3B, 0b10000000) # Gb channel offset (Set to Default 0x80)
sensor.__write_reg(0x3C, 0b10000000) # Gr channel offset (Set to Default 0x80)
```

**Comment by iabdalkader** [comment]
Is there a question/issue here ?

**Comment by kwagyeman** [comment]
I requested this for tracking purposes. I’ll get it done.

**Comment by Niklas123Niklas** [comment]
Hi there,
is this code implemented in the current firmware somehow or do I need to call it manually to gain full manual controll over the camera?
Cheers
Niklas

**Comment by ajacobs-github** [comment]
I don't know if it is in the firmware yet.  Doing it manually with these registers should give full manual control.

**Comment by kwagyeman** [comment]
Full control support has been merged into the main firmware.

**Comment by fatpunkybearass** [comment]
@kwagyeman @ajacobs-github @dhylands @Niklas123Niklas Hi, Im trying this method to manually set exposure time of my ov7725. 
I have some questions:
Is this method using python? I already went a long way using arduino ide, I tried to adapt this code to arduino but failed.  I dont mind transfering to python, but if you guys know how to do it using arduino, I would prefer this way.


Here is my code.


sensor_t* s = esp_camera_sensor_get();
s->set_exposure_ctrl(s, 1);
s->set_aec_value(s, 20);
s->set_aec2(s, 0);
s->get_reg(s,0x0E,0xff);
s->set_reg(s,0x0E,0xff,0); //night mode
s->set_reg(s,0x3E,0xff,0); //BLC
s->set_reg(s,0x2d,0xff,0);
s->set_reg(s,0x2e,0xff,0);
s->set_reg(s,0x00,0xff,0);
s->get_reg(s,0x01,0xff);
s->get_reg(s,0x02,0xff);
s->get_reg(s,0x03,0xff);
s->get_reg(s,0x08,0xff);
s->set_reg(s,0x63,0xff,0x00);
s->set_reg(s,0x10,0xff,0xffff);
s->set_reg(s,0x00,0xff,0x01);
s->set_reg(s,0x08,0xff,0xffff);
s->set_exposure_ctrl(s, 1);
s->set_aec_value(s, 500);
delay(3000);

I found that, whenever i use get_reg, seems every setting of the camera is fixed. 
I start esp32cam in dark, delay 3 seconds, then get_reg. the camera adjusted to high exposure and maintained at this high exposure time no matter the ambient light.  
I start esp32cam in bright, delay 3 seconds, then get_reg. the camera adjusted to low exposure and maintained at this low exposure time no matter the ambient light. 

And basically the set_reg does not work, it still adjusted exposure automatically. I tried many combinations of commands, did not work.

**Comment by kwagyeman** [comment]
The code is here: https://github.com/openmv/openmv/blob/master/src/omv/sensors/ov7725.c#L408

**Comment by fatpunkybearass** [comment]
> The code is here: https://github.com/openmv/openmv/blob/master/src/omv/sensors/ov7725.c#L408

Sorry this arduino is driving me crazy, I have to ask this question: Im not able to include library of this openmv-master.zip. How to include this whole zip file into library?

**Comment by kwagyeman** [comment]
Hi, we don't provide help support for using our code on another system.
